### ISSUE TITLE
0.4: password policy

### ISSUE THREAD

**Issue by webteckie** [keystonejs/keystone#1868]
Should keystone support enforcing a password policy?  Currently I can create a user without a password.

**Comment by ytzvan** [comment]
Add required: true to the passport field in the User Model.

**Comment by webteckie** [comment]
I was more implying an explicit password field type that can configure various password strength rules.

**Comment by webteckie** [comment]
If the current password field has enough to configure a strong password then this may be closed.  I haven't checked what's available.   If people feel that making it be required is enough then another reason to close it.  But while it's open let's consider if there's room for improvement in the current field.

**Comment by mxstbr** [comment]
I agree that this is worth a discussion, https://productpains.com/post/keystonejs/add-min-password-complexity-options would be a good first step

**Comment by bassjacob** [comment]
@webteckie, does the validate hook suffice for this? is there a reason to treat passwords differently than ordinary fields?

As an aside, this might make a good use case for the form validator lib I've been describing. @JedWatson  something along those lines has come up a couple of times now. Is that something we should look into adding explicitly?

**Comment by webteckie** [comment]
@bassjacob I think min/max length support was added at some point after this?  That is an improvement although you right that could have been done with the validate hook as well.  Perhaps if the password field supported a strictness regex expression that may accomplish the rest.  What you think?

**Comment by bassjacob** [comment]
I think rather than setting a custom regex expression validator on the password field we might want to look into a way of setting these as options on all field types? With something like:

```javascript
//...
password: { type: Types.Password, validators: [regex(/^[0-9a-bA-B!@#$%^&*]{9}/)] },
//...

// and then this can be defined wherever
// we can ship some standard ones, and let other people build and compose their own as they please
function regex (pattern) {
  return function validate (input) {
    return new Promise((resolve, reject) => {
      if (!pattern.test(input)) return Promise.reject(new Error(`must match pattern`));
      return Promise.resolve();
    });
  }
}
```

that way we have something a bit more general than a specific pattern protection on a field. what do you think?

**Comment by webteckie** [comment]
Are you referring to the mongoose validator here (as in #2233)  or some validator option in keystone?  By the way @creynders not sure why #2233 was closed.  Last time I checked there were issues reflected in the admin UI when using mongoose validators.  I wanted to make sure the scope of using mongoose validators was analyzed at the keystone admin UI  level to ensure it play well with such validators.

**Comment by VinayaSathyanarayana** [comment]
what is state of this now?
